rls/src/main/java/io/grpc/rls/internal/CachingRlsLbClient.java

```diff
@@ -0,0 +1,979 @@
+          .toString();
+    }
+  }
+
+  /** Returns a Builder for {@link CachingRlsLbClient}. */
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  /** A Builder for {@link CachingRlsLbClient}. */
+  public static final class Builder {
+
+    private Helper helper;
+    private LbPolicyConfiguration lbPolicyConfig;
+    private Throttler throttler = new HappyThrottler();
+    private ResolvedAddressFactory resolvedAddressFactory;
+    private TimeProvider timeProvider = TimeProvider.SYSTEM_TIME_PROVIDER;
+    private EvictionListener<RouteLookupRequest, CacheEntry> evictionListener;
+    private BackoffPolicy.Provider backoffProvider = new ExponentialBackoffPolicy.Provider();
+
+    public Builder setHelper(Helper helper) {
+      this.helper = checkNotNull(helper, "helper");
+      return this;
+    }
+
+    public Builder setLbPolicyConfig(LbPolicyConfiguration lbPolicyConfig) {
+      this.lbPolicyConfig = checkNotNull(lbPolicyConfig, "lbPolicyConfig");
+      return this;
+    }
+
+    public Builder setThrottler(Throttler throttler) {
+      this.throttler = checkNotNull(throttler, "throttler");
+      return this;
+    }
+
+    /**
+     * Sets a factory to create {@link ResolvedAddresses} for child load balancer.
+     */
+    public Builder setResolvedAddressesFactory(
+        ResolvedAddressFactory resolvedAddressFactory) {
+      this.resolvedAddressFactory =
+          checkNotNull(resolvedAddressFactory, "resolvedAddressFactory");
+      return this;
+    }
+
+    public Builder setTimeProvider(TimeProvider timeProvider) {
+      this.timeProvider = checkNotNull(timeProvider, "timeProvider");
+      return this;
+    }
+
+    public Builder setEvictionListener(
+        @Nullable EvictionListener<RouteLookupRequest, CacheEntry> evictionListener) {
+      this.evictionListener = evictionListener;
+      return this;
+    }
+
+    public Builder setBackoffProvider(BackoffPolicy.Provider provider) {
+      this.backoffProvider = checkNotNull(provider, "provider");
+      return this;
+    }
+
+    public CachingRlsLbClient build() {
+      return new CachingRlsLbClient(this);
+    }
+  }
+
+  /**
+   * When any {@link CacheEntry} is evicted from {@link LruCache}, it performs {@link
+   * CacheEntry#cleanup()} after original {@link EvictionListener} is finished.
+   */
+  private static final class AutoCleaningEvictionListener
+      implements EvictionListener<RouteLookupRequest, CacheEntry> {
+
+    private final EvictionListener<RouteLookupRequest, CacheEntry> delegate;
+
+    AutoCleaningEvictionListener(
+        @Nullable EvictionListener<RouteLookupRequest, CacheEntry> delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public void onEviction(RouteLookupRequest key, CacheEntry value, EvictionType cause) {
+      if (delegate != null) {
+        delegate.onEviction(key, value, cause);
+      }
+      // performs cleanup after delegation
+      value.cleanup();
```

Comment:
`LinkedHashLruCache` is guarded by `CachingRlsLbClient.lock`.
`LinkedHashLruCache.evictionListener.onEviction()` is guarded by `LinkedHashLruCache.lock`.
There could be deadlock:

`LinkedHashLruCache.cache() -> evictionListener.onEviction() -> value.cleanup()`

Reply:
oh yes good catch. the lock in cleanup is not required any more. it was added when I was using invalidate with key value. i am glad  your previous finding of the invalidate(Key, Value), it fixed two issues =p.